await session.interrupt() with gpt-realtime model not reliable

### Bug Description


We have code in our agent which triggers speech with gpt-realtime via session.generate_reply() when certain events happen, e.g. user speaking too long etc.

When using livekit testing framework (text only mode) we noticed that we could not test the functionality well because it seems `await interrupt()` is not cancelling in flight requests before subequent requests, meaning OpenAI rejects subsequent requests (` RealtimeError(message='Conversation already has an active response in progress: resp_Dc7WDsAfSkZTNf34lhYGX. Wait until the response is finished before creating a new one.'`)

I have provided a minimal reproducible example of this behaviour on this issue.

I think that this issue _might_  not be present in production (outside of text only tests) because in real usage, the interrupt awaits the speech handle stopping, so in practice would correctly wait until the interrupt has finished in most cases (as interrupt event should trigger speech stopping). Nonetheless I think there is value in fixing this becuase:
a) any divergence between the test framework and production in behaviour is problematic for testing. 
b) I am not sure that speech handle ending would be 100% reliable indicator that the interrupt has been successfully completed in all possible scenarios (e.g. what about tool call cancellation?) 

### Expected Behavior

- await interrupt() whould await OpenAI ackowledgement that interrupt has been received
- The attached test would pass 


### Reproduction Steps

```bash
import asyncio
import logging

import pytest
from dotenv import load_dotenv
from livekit.agents import Agent, AgentSession
from livekit.plugins import openai
from openai.types.realtime import AudioTranscription
from openai.types.realtime.realtime_audio_input_turn_detection import SemanticVad

load_dotenv(".env.local")

_REALTIME_ACTIVE_RESPONSE_SNIPPET = "conversation_already_has_active_response"


class Assistant(Agent):
    """Minimal agent for interrupt / follow-up turn tests."""

    def __init__(self) -> None:
        super().__init__(
            instructions="You are a helpful voice AI assistant. Speak English.",
        )


def _semantic_vad() -> SemanticVad:
    return SemanticVad(
        type="semantic_vad",
        create_response=True,
        interrupt_response=True,
        eagerness="low",
    )


def _realtime_llm() -> openai.realtime.RealtimeModel:
    return openai.realtime.RealtimeModel(
        modalities=["text"],
        input_audio_transcription=AudioTranscription(language="en", model="whisper-1"),
        max_session_duration=20 * 60,
        turn_detection=_semantic_vad(),
    )


@pytest.mark.asyncio
async def test_interrupt_then_follow_up_user_turn(caplog: pytest.LogCaptureFixture) -> None:
    """User-driven turns; interrupt mid first reply; second reply reflects the new ask."""
    with caplog.at_level(logging.ERROR, logger="livekit.plugins.openai"):
        async with (
            _realtime_llm() as realtime_llm,
            AgentSession(llm=realtime_llm, turn_detection="realtime_llm") as session,
        ):
            await session.start(Assistant())

            session.generate_reply(user_input="Reply with 5 sentence essay about cats")
            await asyncio.sleep(0.1)
            await session.interrupt()

            second = session.run(
                user_input=(
                    "Ignore any partial earlier reply. Reply with only the exact phrase "
                    "I HATE DOGS in all caps. Do not add any other words."
                )
            )
            await second

    error_messages = [r.getMessage() for r in caplog.records if r.levelno >= logging.ERROR]
    assert not any(_REALTIME_ACTIVE_RESPONSE_SNIPPET in m for m in error_messages), (
        "OpenAI Realtime logged active-response conflict "
        f"({_REALTIME_ACTIVE_RESPONSE_SNIPPET}): {error_messages!r}"
    )


````
Fails with 


E   AssertionError: OpenAI Realtime logged active-response conflict (conversation_already_has_active_response): ["OpenAI Realtime API returned an error: RealtimeError(message='Conversation already has an active response in progress: resp_Dc7WDsAfSkZTNf34lhYGX. Wait until the response is finished before creating a new one.', type='invalid_request_error', code='conversation_already_has_active_response', event_id='response_create_67056e52992a', param=None)", 'failed to generate a reply: generate_reply timed out.']

### Operating System

Linux

### Models Used

gpt-realtime

### Package Versions

```bash
1.5.7
```

### Session/Room/Call IDs

_No response_

### Proposed Solution

```python
For gpt-realtime, await interrupt() should wait for the OpenAI `cancelled 'acknowledgement' event to be received (with a timeout in case it never is received).
```

### Additional Context

_No response_

### Screenshots and Recordings

_No response_

--- a/livekit-agents/livekit/agents/voice/agent_activity.py
+++ b/livekit-agents/livekit/agents/voice/agent_activity.py
@@ -2888,18 +2888,24 @@ class AgentActivity(RecognitionHooks):
                     ori_tools = self._rt_session.tools.flatten()
                     await self._rt_session.update_tools(llm.ToolContext(tools).flatten())
 
+            generate_reply_fut = self._rt_session.generate_reply(
+                instructions=instructions or NOT_GIVEN,
+                tool_choice=(model_settings.tool_choice if per_response_tool_choice else NOT_GIVEN),
+                tools=(
+                    llm.ToolContext(tools).flatten()
+                    if per_response_tool_choice and tools is not None
+                    else NOT_GIVEN
+                ),
+            )
+            await speech_handle.wait_if_not_interrupted([generate_reply_fut])
+            if speech_handle.interrupted:
+                # cancel the pending generation; the plugin emits response.cancel
+                if not generate_reply_fut.done():
+                    generate_reply_fut.cancel()
+                return
+
             try:
-                generation_ev = await self._rt_session.generate_reply(
-                    instructions=instructions or NOT_GIVEN,
-                    tool_choice=(
-                        model_settings.tool_choice if per_response_tool_choice else NOT_GIVEN
-                    ),
-                    tools=(
-                        llm.ToolContext(tools).flatten()
-                        if per_response_tool_choice and tools is not None
-                        else NOT_GIVEN
-                    ),
-                )
+                generation_ev = generate_reply_fut.result()
             except llm.RealtimeError as e:
                 logger.error(
                     "failed to generate a reply%s: %s",

--- a/livekit-agents/livekit/agents/voice/agent_activity.py
+++ b/livekit-agents/livekit/agents/voice/agent_activity.py
@@ -2905,7 +2905,7 @@ class AgentActivity(RecognitionHooks):
                 return
 
             try:
-                generation_ev = generate_reply_fut.result()
+                generation_ev = await generate_reply_fut
             except llm.RealtimeError as e:
                 logger.error(
                     "failed to generate a reply%s: %s",

--- a/livekit-plugins/livekit-plugins-aws/livekit/plugins/aws/experimental/realtime/realtime_model.py
+++ b/livekit-plugins/livekit-plugins-aws/livekit/plugins/aws/experimental/realtime/realtime_model.py
@@ -2116,7 +2116,7 @@ class RealtimeSession(  # noqa: F811
                     if self._pending_generation_fut is fut:
                         self._pending_generation_fut = None
 
-            asyncio.create_task(_send_text())
+            send_task = asyncio.create_task(_send_text())
 
             # Set timeout from model configuration
             def _on_timeout() -> None:
@@ -2130,7 +2130,17 @@ class RealtimeSession(  # noqa: F811
             timeout_handle = asyncio.get_running_loop().call_later(
                 self._realtime_model._generate_reply_timeout, _on_timeout
             )
-            fut.add_done_callback(lambda _: timeout_handle.cancel())
+
+            def _on_fut_done(f: asyncio.Future[llm.GenerationCreatedEvent]) -> None:
+                timeout_handle.cancel()
+                is_current = self._pending_generation_fut is fut
+                if is_current:
+                    self._pending_generation_fut = None
+                if f.cancelled() and is_current and not send_task.done():
+                    # external cancel before the text was sent: drop the send
+                    send_task.cancel()
+
+            fut.add_done_callback(_on_fut_done)
 
             return fut
 

--- a/livekit-plugins/livekit-plugins-google/livekit/plugins/google/realtime/realtime_api.py
+++ b/livekit-plugins/livekit-plugins-google/livekit/plugins/google/realtime/realtime_api.py
@@ -717,7 +717,11 @@ class RealtimeSession(llm.RealtimeSession):
             logger.warning(
                 "generate_reply called while another generation is pending, cancelling previous."
             )
-            self._pending_generation_fut.cancel("Superseded by new generate_reply call")
+            # clear the slot before cancelling so the done callback doesn't treat it
+            # as an external cancellation and signal the server.
+            old_fut = self._pending_generation_fut
+            self._pending_generation_fut = None
+            old_fut.cancel("Superseded by new generate_reply call")
 
         fut = asyncio.Future[llm.GenerationCreatedEvent]()
         self._pending_generation_fut = fut
@@ -749,7 +753,17 @@ class RealtimeSession(llm.RealtimeSession):
                     self._pending_generation_fut = None
 
         timeout_handle = asyncio.get_event_loop().call_later(5.0, _on_timeout)
-        fut.add_done_callback(lambda _: timeout_handle.cancel())
+
+        def _on_fut_done(f: asyncio.Future[llm.GenerationCreatedEvent]) -> None:
+            timeout_handle.cancel()
+            is_current = self._pending_generation_fut is fut
+            if is_current:
+                self._pending_generation_fut = None
+            if f.cancelled() and is_current:
+                # external cancel: signal interrupt to Gemini via activity_start
+                self.interrupt()
+
+        fut.add_done_callback(_on_fut_done)
 
         return fut
 

--- a/livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/realtime/realtime_model.py
+++ b/livekit-plugins/livekit-plugins-openai/livekit/plugins/openai/realtime/realtime_model.py
@@ -1532,7 +1532,15 @@ class RealtimeSession(
                 fut.set_exception(llm.RealtimeError("generate_reply timed out."))
 
         handle = asyncio.get_event_loop().call_later(10.0, _on_timeout)
-        fut.add_done_callback(lambda _: handle.cancel())
+
+        def _on_fut_done(f: asyncio.Future[llm.GenerationCreatedEvent]) -> None:
+            handle.cancel()
+            self._response_created_futures.pop(event_id, None)
+            if f.cancelled():
+                # response.create was already sent; cancel the response server-side
+                self.send_event(ResponseCancelEvent(type="response.cancel"))
+
+        fut.add_done_callback(_on_fut_done)
         return fut
 
     @property

--- a/livekit-plugins/livekit-plugins-phonic/livekit/plugins/phonic/realtime/realtime_model.py
+++ b/livekit-plugins/livekit-plugins-phonic/livekit/plugins/phonic/realtime/realtime_model.py
@@ -474,12 +474,17 @@ class RealtimeSession(llm.RealtimeSession):
         )
         if self._generate_reply_task and not self._generate_reply_task.done():
             self._generate_reply_task.cancel()
-        self._generate_reply_task = asyncio.create_task(self._send_generate_reply(payload))
+        send_task = asyncio.create_task(self._send_generate_reply(payload))
+        self._generate_reply_task = send_task
 
         self._close_current_generation(interrupted=False)
 
         if self._pending_generate_reply_fut and not self._pending_generate_reply_fut.done():
-            self._pending_generate_reply_fut.cancel()
+            # clear the slot first so the done callback doesn't see this as an
+            # external cancellation of the currently-pending generation.
+            old_fut = self._pending_generate_reply_fut
+            self._pending_generate_reply_fut = None
+            old_fut.cancel()
 
         fut = asyncio.Future[llm.GenerationCreatedEvent]()
         self._pending_generate_reply_fut = fut
@@ -489,7 +494,18 @@ class RealtimeSession(llm.RealtimeSession):
                 fut.set_exception(llm.RealtimeError("generate_reply timed out."))
 
         handle = asyncio.get_event_loop().call_later(10.0, _on_timeout)
-        fut.add_done_callback(lambda _: handle.cancel())
+
+        def _on_fut_done(f: asyncio.Future[llm.GenerationCreatedEvent]) -> None:
+            handle.cancel()
+            is_current = self._pending_generate_reply_fut is fut
+            if is_current:
+                self._pending_generate_reply_fut = None
+            if f.cancelled() and is_current:
+                # external cancel: drop the queued send if it hasn't gone out yet
+                if not send_task.done():
+                    send_task.cancel()
+
+        fut.add_done_callback(_on_fut_done)
         return fut
 
     async def _send_generate_reply(self, payload: GenerateReplyPayload) -> None:

--- a/livekit-plugins/livekit-plugins-ultravox/livekit/plugins/ultravox/realtime/realtime_model.py
+++ b/livekit-plugins/livekit-plugins-ultravox/livekit/plugins/ultravox/realtime/realtime_model.py
@@ -490,7 +490,11 @@ class RealtimeSession(
             logger.warning(
                 "generate_reply called while another generation is pending, cancelling previous."
             )
-            self._pending_generation_fut.cancel("Superseded by new generate_reply call")
+            # clear the slot before cancelling so the done callback doesn't treat it
+            # as an external cancellation and signal the server.
+            old_fut = self._pending_generation_fut
+            self._pending_generation_fut = None
+            old_fut.cancel("Superseded by new generate_reply call")
 
         # Record epoch for server-event gating
         self._pending_generation_epoch = time.perf_counter()
@@ -522,7 +526,21 @@ class RealtimeSession(
                     self._pending_generation_epoch = None
 
         timeout_handle = asyncio.get_event_loop().call_later(5.0, _on_timeout)
-        fut.add_done_callback(lambda _: timeout_handle.cancel())
+
+        def _on_fut_done(f: asyncio.Future[llm.GenerationCreatedEvent]) -> None:
+            timeout_handle.cancel()
+            is_current = self._pending_generation_fut is fut
+            if is_current:
+                self._pending_generation_fut = None
+                self._pending_generation_epoch = None
+            if f.cancelled() and is_current:
+                # external cancel: send a deferred barge-in so Ultravox does not
+                # produce a response (or interrupts one already in progress).
+                self._send_client_event(
+                    UserTextMessageEvent(text="", urgency="immediate", defer_response=True)
+                )
+
+        fut.add_done_callback(_on_fut_done)
 
         return fut
 

--- a/tests/test_realtime/test_realtime.py
+++ b/tests/test_realtime/test_realtime.py
@@ -312,6 +312,28 @@ async def test_interrupt(rt_session: llm.RealtimeSession):
     assert got_chunk
 
 
+@pytest.mark.parametrize("rt_session", OPENAI_AND_AZURE, indirect=True)
+async def test_generate_reply_cancellation(rt_session: llm.RealtimeSession):
+    """Cancelling the generate_reply future before response.created arrives
+    should cancel the in-flight response server-side so a subsequent
+    generate_reply does not conflict with an active response."""
+    fut = rt_session.generate_reply(
+        instructions="Write a very long essay about the history of computing."
+    )
+    fut.cancel()
+    assert fut.cancelled()
+
+    # Brief wait so the response.cancel reaches the server before the next call.
+    await asyncio.sleep(1.0)
+
+    gen_ev = await asyncio.wait_for(
+        rt_session.generate_reply(instructions="Say exactly: pineapple"),
+        timeout=15,
+    )
+    text = await asyncio.wait_for(_collect_text(gen_ev), timeout=15)
+    assert "pineapple" in text.lower()
+
+
 # -- Function tools --
 
 

